File: LeetCodeProblemsCMake/Problems/LeetCode1353MaximumNumberOfEventsThatCanBeAttended.cpp
//
// Created by saeed on 7/7/25.
//

#include "LeetCode1353MaximumNumberOfEventsThatCanBeAttended.h"

#include <algorithm>
#include <iostream>
#include <ostream>
#include <queue>


using namespace std;

int LeetCode1353MaximumNumberOfEventsThatCanBeAttended::maxEvents(vector<vector<int> > &events) {
    ranges::sort(events, [](const vector<int> &a, const vector<int> &b) {
        if (a[0] == b[0]) {
            return a[1] < b[1];
        }
        return a[0] < b[0];
    });
    auto cmp = [](const vector<int> &a, const vector<int> &b) {
        return a[1] > b[1];
    };
    priority_queue<vector<int>, vector<vector<int> >, decltype(cmp)> minHeap(cmp);

    auto length = events.size();
    auto maxDay = events[length - 1][1];

    int eventsCovered{0};
    int j{0};
    for (int i = events[0][0]; i <= maxDay; i++) {
        while (j < length && events[j][0] <= i) {
            if (events[j][1] > maxDay) {
                maxDay = events[j][1];
            }
            minHeap.push(events[j]);
            j++;
        }
        while (!minHeap.empty() && minHeap.top()[1] < i) {
            minHeap.pop();
        }
        if(minHeap.empty())
          continue;
        auto event = minHeap.top();
        minHeap.pop();
        if (i >= event[0] && i <= event[1]) {
            eventsCovered++;
        }
    }

    return eventsCovered;
}

void LeetCode1353MaximumNumberOfEventsThatCanBeAttended::Run() {
    vector<vector<int> > events =
            // {{1, 5}, {1, 5}, {1, 5}, {2, 3}, {2, 3}};
            {{1, 2}, {1, 2}, {3, 3}, {1, 5}, {1, 5}};
    // {{1, 4}, {4, 4}, {2, 2}, {3, 4}, {1, 1}};
    // {{1,2},{1,2},{1,6},{1,2},{1,2}};
    cout << maxEvents(events) << endl;
}

File: LeetCodeProblemsGo/LeetCode1353MaximumNumberOfEventsThatCanBeAttended/LeetCode1353MaximumNumberOfEventsThatCanBeAttended.go
package LeetCode1353MaximumNumberOfEventsThatCanBeAttended

import (
	"container/heap"
	"sort"
)

func maxEvents(events [][]int) int {
	sort.Slice(events, func(i, j int) bool {
		if events[i][0] == events[j][0] {
			return events[i][1] < events[j][1]
		}
		return events[i][0] < events[j][0]
	})
	length := len(events)
	maxDay := events[length-1][1]
	minHeap := &minHeap{}
	heap.Init(minHeap)
	coveredEvents := 0
	eventIndex := 0
	for day := 1; day <= maxDay; day++ {
		for eventIndex < length && events[eventIndex][0] <= day {
			if events[eventIndex][1] > maxDay {
				maxDay = events[eventIndex][1]
			}
			heap.Push(minHeap, events[eventIndex])
			eventIndex++
		}
		for !minHeap.IsEmpty() && minHeap.Top()[1] < day {
			heap.Pop(minHeap)
		}
		if !minHeap.IsEmpty() {
			heap.Pop(minHeap)
			coveredEvents++
		}

	}
	return coveredEvents
}

type minHeap [][]int

func (h *minHeap) Len() int {
	return len(*h)
}
func (h *minHeap) Less(i, j int) bool {
	return (*h)[i][1] < (*h)[j][1]
}
func (h *minHeap) Swap(i, j int) {
	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
}
func (h *minHeap) Push(x any) {
	*h = append(*h, x.([]int))
}
func (h *minHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}
func (h *minHeap) Top() []int {
	if len(*h) == 0 {
		return nil
	}
	return (*h)[0]
}
func (h *minHeap) IsEmpty() bool {
	return len(*h) == 0
}
